lib/web_ui/lib/src/engine/initialization.dart

```diff
@@ -144,8 +215,93 @@ void initializeEngine() {
     }
   };
 
+  // This needs to be after `initializeEngine` because that is where the
+  // canvaskit script is added to the page.
+  if (useCanvasKit) {
+    await initializeCanvasKit();
+  }
+
+  assetManager ??= const AssetManager();
+  await _setAssetManager(assetManager);
+  if (useCanvasKit) {
+    await skiaFontCollection.ensureFontsLoaded();
+  } else {
+    await _fontCollection!.ensureFontsLoaded();
+  }
+  _initializationState = DebugEngineInitializationState.initializedServices;
+}
+
+/// Initializes the UI surfaces for the Flutter framework to render to.
+///
+/// Must be called after [initializeEngineServices].
+///
+/// This function will start altering the HTML structure of the page. If used
+/// in an add-to-app scenario, the host page is expected to prepare for Flutter
+/// UI appearing on screen prior to calling this function.
+Future<void> initializeEngineUi() async {
+  if (_initializationState != DebugEngineInitializationState.initializedServices) {
+    assert(() {
+      throw StateError(
+        'Invalid engine initialization state. `initializeEngineUi` was '
+        'called while the engine initialization state was '
+        '"$_initializationState". `initializeEngineUi` can only be called '
+        'when the engine is in state '
+        '"${DebugEngineInitializationState.initializedServices}".'
+      );
+    }());
+    return;
+  }
+  _initializationState = DebugEngineInitializationState.initializingUi;
+
   Keyboard.initialize(onMacOs: operatingSystem == OperatingSystem.macOs);
   MouseCursor.initialize();
+  ensureFlutterViewEmbedderInitialized();
+
+  if (useCanvasKit) {
+    /// Add a Skia scene host.
+    skiaSceneHost = html.Element.tag('flt-scene');
+    flutterViewEmbedder.renderScene(skiaSceneHost);
+  }
+  _initializationState = DebugEngineInitializationState.initialized;
+}
+
+AssetManager get assetManager => _assetManager!;
+AssetManager? _assetManager;
+
+FontCollection get fontCollection => _fontCollection!;
+FontCollection? _fontCollection;
+
+Future<void> _setAssetManager(AssetManager assetManager) async {
+  // ignore: unnecessary_null_comparison
+  assert(assetManager != null, 'Cannot set assetManager to null');
```

Comment:
Is this null check still needed?

Reply:
Yes. While Flutter itself migrated to null safety, apps/plugins/packages may not have, so we can still see a null passed here.

Reply:
I supposed so. I guess we need to assume anything passed in from JS may be `?` for later changes.

Reply:
Ah yeah, JS will never be sound, so we may need to keep some checks even when all Dart code is 100% migrated. Well, at least until we have dart2wasm